swap/swap.go

```diff
@@ -143,28 +145,60 @@ func New(logpath string, dbPath string, prvkey *ecdsa.PrivateKey, backendURL str
 	// initialize the balances store
 	stateStore, err := state.NewDBStore(filepath.Join(dbPath, "swap.db"))
 	if err != nil {
-		return nil, fmt.Errorf("swap init error: %s", err)
+		return nil, fmt.Errorf("swap init error: error initializing statestore: %v", err)
 	}
 	if disconnectThreshold <= paymentThreshold {
 		return nil, fmt.Errorf("swap init error: disconnect threshold lower or at payment threshold. DisconnectThreshold: %d, PaymentThreshold: %d", disconnectThreshold, paymentThreshold)
 	}
 	backend, err := ethclient.Dial(backendURL)
 	if err != nil {
-		return nil, fmt.Errorf("swap init error: error connecting to Ethereum API %s: %s", backendURL, err)
+		return nil, fmt.Errorf("swap init error: error connecting to Ethereum API %s: %v", backendURL, err)
 	}
+	networkID, err := backend.NetworkID(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("swap init error: error retrieving networkID from backendURL: %v", err)
+	}
+	var usedBeforeAtNetwork uint64
+	var usedBefore bool
+	err = stateStore.Get(usedBeforeAtNetworkKey, &usedBeforeAtNetwork)
+	// error reading from database
+	if err != nil && err != state.ErrNotFound {
+		return nil, fmt.Errorf("swap init error: error querying usedBeforeAtNetwork from statestore: %v", err)
+	}
+	// read from database
+	if err == nil {
+		usedBefore = true
+	}
+	// if used before on a different network, abort
+	if usedBefore && usedBeforeAtNetwork != networkID.Uint64() {
+		return nil, fmt.Errorf("swap init error: statestore previously used on different backend network. Used before on network: %d, Attempting to connect on network %d", usedBeforeAtNetwork, networkID)
+	}
+	// put the networkID in the statestore
+	if !usedBefore {
+		err = stateStore.Put(usedBeforeAtNetworkKey, networkID.Uint64())
+		if err != nil {
+			return nil, fmt.Errorf("swap init error: error writing current backend networkID to statestore: %v", err)
+		}
+		log.Info("SWAP initialized on backend network ID", "ID", networkID.Uint64())
+	}
+
 	return new(
 		logpath,
 		stateStore,
 		prvkey,
 		backend,
 		disconnectThreshold,
-		paymentThreshold), nil
+		paymentThreshold,
+		networkID.Uint64(),
+	), nil
```

Comment:
I agree, this whole initialization process is long and not very readable.

I also think it could be compressed. Not well verified but you could check if this works (and if it is correct):
```
if err == nil {
  if  usedBeforeAtNetwork != network.Uint64() {
  ...
   }
} else {
   stateStore.Put()
}
...

Maybe there is even a more elegant way.

Reply:
I put the logic in a separate function